mono/utils/mono-threads-windows.c

```diff
@@ -289,6 +289,49 @@ mono_threads_platform_get_stack_bounds (guint8 **staddr, size_t *stsize)
 
 }
 
+typedef BOOL (WINAPI *LPFN_ISWOW64PROCESS) (HANDLE, PBOOL);
+static gboolean is_wow64 = FALSE;
+
+/* We do this at init time to avoid potential races with module opening */
+void
+mono_threads_platform_init (void)
+{
+	LPFN_ISWOW64PROCESS is_wow64_func = (LPFN_ISWOW64PROCESS) GetProcAddress (GetModuleHandle (TEXT ("kernel32")), "IsWow64Process");
+	if (is_wow64_func)
+		is_wow64_func (GetCurrentProcess (), &is_wow64);
+}
+
+/*
+ * When running x86 process under x64 system syscalls are done through WoW64. This
+ * needs to do a transition from x86 mode to x64 so it can syscall into the x64 system.
+ * Apparently this transition invalidates the ESP that we would get from calling
+ * GetThreadContext, so we would fail to scan parts of the thread stack. We attempt
+ * to query whether the thread is in such a transition so we try to restart it later.
+ * We check CONTEXT_EXCEPTION_ACTIVE for this, which is highly undocumented.
+ */
+gboolean
+mono_threads_platform_in_critical_region (MonoNativeThreadId tid)
+{
+	gboolean ret = FALSE;
+#if defined(CONTEXT_EXCEPTION_ACTIVE) && defined(CONTEXT_EXCEPTION_REQUEST) && defined(CONTEXT_EXCEPTION_REPORTING)
+	if (is_wow64) {
+		HANDLE handle = OpenThread (THREAD_ALL_ACCESS, FALSE, tid);
+		if (handle) {
+			CONTEXT context;
+			ZeroMemory (&context, sizeof (CONTEXT));
+			context.ContextFlags = CONTEXT_EXCEPTION_REQUEST;
+			if (GetThreadContext (handle, &context)) {
+				if ((context.ContextFlags & CONTEXT_EXCEPTION_REPORTING) &&
+						(context.ContextFlags & CONTEXT_EXCEPTION_ACTIVE))
+					ret = TRUE;
+			}
+			CloseHandle (handle);
+		}
+	}
+#endif
+	return ret;
+}
+
```

Comment:
This section is only needed for x86 so there should not be a need to include the code in x64 windows builds. Since the code also uses API's that are not present on all Windows API families, I believe we should just start to include it for windows desktop. That can be achieved by using the following define:

#if G_HAVE_API_SUPPORT(HAVE_CLASSIC_WINAPI_SUPPORT)